Enterprise-CMCS/eAPD
# Ad/4211 MMIS APD overview

Resolves #4211 

### Description
- Updating APD Overview to allow an MMIS APD type. This includes removing fields specific to HITECH and adding a field specific to MMIS
- Addition of APD Update field to APD Overview page
- Adds validation to APD Name

### Significant changes or possible side effects
- Schema updates:
  - MMIS medicaidBusinessAreas tweaking for validation purposes
  - APD Name schema includes ensuring the name does not contain "untitled"
  - APD Update schema created for each of MMIS and HITECH since the validations differ
- Redux action additions:
  - action to update a key value pair in the `medicaidBusinessAreas` object
  - action to update a key value pair in the `updateStatus` object
- Component changes:
  - Set up APD Name in `ApdOverview` to be handled within react hook form for validation purposes
  - Modified `ApdOverview` to contain only the shared fields of HITECH and MMIS APDs. Subsequently moved HITECH fields to new component called `ApdOverviewHITECHFields` and created `ApdOverviewMMISFields` for the MMIS fields.
  - Created `ApdUpdate` component to handle displaying the Update APD section differently for MMIS and HITECH types
  - Added `disableTitleCaseConversion` prop to the `Instruction` component
- Backend
  -  Added APD Name validation to fall under APD Overview in Admin Check
- Misc
  - Created a constant for mapping Medicaid Business Areas key names to labels
  - Updated seed data for invalid and valid options regarding updateStatus field
  - Updated label sizing for fields on APD Overview page
  - Resolved some existing console warnings for "APD Type" field by adding `name` and `label` props. Since the code was written with the intent of the label being hidden, a `labelClassName` prop was also passed in to include a newly added `hidden-display` style class which hides the label

### Automated test cases written

| Given | When | Then | Type (jest, tap, cypress) |
| ----- | ---- | ---- | ------------------------- |
APD Name with Admin Check ON | Name contains "untitled" | Validation error appears on the field and in admin check | Cypress
APD Name | Previous data exists for name | Displays previous data in the APD Name field | Jest
MMIS APD Overview medicaid business areas  | No option selected | Validation error appears on the field and in admin check | Cypress
MMIS APD Overview medicaid business areas  | "Other" option selected, but "Other Medicaid Business Areas" text box is empty | Validation error appears on the field and in admin check | Cypress
MMIS APD Overview medicaid business areas with Admin Check ON | Form errors are active | Validation error also appears in admin check | Cypress
MMIS Medicaid Business Areas | Previous data exists | Displays checkbox as selected or unselected based on existing data | Jest
MMIS Medicaid Business Areas | "Other" option is checked | "Other Medicaid Business Area(s)" text area and content displays | Jest
MMIS Medicaid Business Areas | "Other" option is NOT checked | "Other Medicaid Business Area(s)" text area is NOT visible | Jest
HITECH APD Overview | Each of the text fields | Allows text changes | Jest
APD Update in APD Overview | HITECH type | Displays "Update Type" checkbox options | Jest & Cypress
APD Update in APD Overview | MMIS type | Displays "Is this an APD update?" radio options | Jest & Cypress
MMIS APD Overview APD Update field | "Yes, it is an update." radio option is selected | "Update Type" options available | Jest & Cypress
APD Overview APD Update field with Admin Check ON | "Update Type" options are all unchecked | Validation error appears on the field and in admin check | Cypress
Instruction component | default behavior for heading | heading is converted to title case | Jest
Instruction component | `disableTitleCaseConversion` is passed in | original heading text appears as is (NOT converted to title case) | Jest
APD edit actions for Medicaid Business Areas | `setBusinessAreaField` is called | redux dispatches an action for changing the field value | unit test

### Steps to verify changes
_With admin check ON_
1. Open an existing MMIS APD
1. Go to APD Overview
2. Test APD Name
    1. Modify the APD name (try including "untitled" in the name) to ensure that errors display both on the form AND in admin check
    4. Modify the APD name _from the APD Header_ and ensure that the name changes reflect within the APD Overview form and that validation errors accurately change on the form.
5. Test APD Update 
    1. Select "Yes, it is an update."
    2. Ensure that the "Update Type" section only appears when "Yes, it is an update." is selected.
    7. Ensure error to "Select an update type." appears in admin check and on the form.
    8. Ensure error disappears when either an update type is selected or the radio button choice is switched to "No, this is for a new project."
1. Test Medicaid Business Areas
    1. Unselect any checked Medicaid Business Areas to ensure the validation error requiring a selection display both on the form AND in admin check
    11. Select "Other" checkbox in Medicaid Business Areas to ensure a text box appears for "Other Medicaid Business Areas". Check that validation error for text box displays when empty (both on the form AND in admin check)

_With admin check OFF_
1. Go to an existing MMIS APD
3. Ensure that none of the validation errors appear on the form

_Other testing_
1. HITECH APD Overview
    1. Regression testing
    3. Update Type field displays
    4. _With admin check on,_ ensure error to "Select an update type." appears in admin check and on the form


### This pull request is ready to code review when

      tests, refactor
      screenreader, text scaling) OR an exemption is documented

### This pull request is ready to test when


### This pull request is ready to review when the QA has


### This pull request can be merged when
